service/operator/service.go

```diff
@@ -163,116 +159,331 @@ func (s *Service) addFunc(obj interface{}) {
+		}
 	}
 
-	clusterPort, err := s.newClusterPort(k8sService.Spec.Ports, ipsPortsToPorts(s.ipsPorts))
-	if err != nil {
-		return microerror.MaskAny(err)
+	// Lookup the desired state of the service to have a reference of ports how
+	// they should be.
+	{
+		for _, p := range customObject.Spec.ProtocolPorts {
+			servicePortName := fmt.Sprintf(
+				ServicePortNameFormat,
+				p.Protocol,
+				p.IngressPort,
+				customObject.Spec.GuestCluster.ID,
+			)
+
+			newPort := apiv1.ServicePort{
+				Name:       servicePortName,
+				Protocol:   apiv1.ProtocolTCP,
+				Port:       int32(p.LBPort),
+				TargetPort: intstr.FromInt(p.LBPort),
+				NodePort:   int32(p.LBPort),
+			}
+
+			dState.ServicePorts = append(dState.ServicePorts, newPort)
+		}
+	}
+
+	return dState, nil
+}
+
+func (s *Service) GetEmptyState() interface{} {
+	return DesiredState{}
+}
+
+func (s *Service) GetCreateState(obj, currentState, desiredState interface{}) (interface{}, error) {
+	customObject, ok := obj.(*ingresstpr.CustomObject)
+	if !ok {
+		return nil, microerror.MaskAnyf(wrongTypeError, "expected '%T', got '%T'", &ingresstpr.CustomObject{}, obj)
+	}
+	cState, ok := currentState.(CurrentState)
+	if !ok {
+		return nil, microerror.MaskAnyf(wrongTypeError, "expected '%T', got '%T'", CurrentState{}, currentState)
+	}
+	dState, ok := desiredState.(DesiredState)
+	if !ok {
+		return nil, microerror.MaskAnyf(wrongTypeError, "expected '%T', got '%T'", DesiredState{}, desiredState)
 	}
 
-	newPort := apiv1.ServicePort{
-		Name:       portName,
-		Protocol:   apiv1.ProtocolTCP,
-		Port:       int32(clusterPort),
-		TargetPort: intstr.FromInt(clusterPort),
-		NodePort:   int32(clusterPort),
+	s.logger.Log("debug", "get create state", "cluster", customObject.Spec.GuestCluster.ID)
+
+	var createState ActionState
+
+	// Make sure the current state of the Kubernetes resources is known by the
+	// create action. The resources we already fetched represent the source of
+	// truth. They have to be used as base to actually update the resources in the
+	// next steps.
+	{
+		createState.ConfigMap = cState.ConfigMap
+		createState.Service = cState.Service
 	}
 
-	k8sService.Spec.Ports = append(k8sService.Spec.Ports, newPort)
+	// Find anything which is in desired state but not in the current state. This
+	// lets us drive the current state towards the desired state, because
+	// everything we find here is supposed to be created. Note that the creation
+	// of config-map data and service ports is always only an update operation
+	// against the Kubernetes API. Anyway, this concept here implements how a real
+	// reconciliation should drive specific parts of the current state towards the
+	// desired state, because a decent reconciliation is not always only an update
+	// operation of existing resources, but e.g. creation of resources. In our
+	// case here we only transform data within resources. Therefore the update.
+	{
+		// Process config-map to find its create state.
+		{
```

Comment:
Just cosmetic to make me happy. The code is more structured that way. In general it is a code smell that indicates the design is not sufficient. This is on purpose now and I do it often to highlight which code parts logically belong together.